### PR TITLE
add a ServiceMonitor to scrape RedPanda metrics

### LINKED ISSUE #149: support prometheus servicemonitors
it would be great if the chart would support servicemonitors for prometheus.
this would involve two things:

- create a service monitor template
- add back the `admin` port in the `service.internal.yaml` as the servicemonitor targets `/metrics` path at the admin port

### PR DESCRIPTION
**Pull request by prune998**
Fixes #149

### CODE REVIEW

File: charts/redpanda/templates/service.internal.yaml [hunk 1]
```diff
@@ -16,14 +16,15 @@ limitations under the License.
...
-# application
+# the stateful set and allow the serviceMonitor to target the pods.
+# This service should not be used by any client application
 apiVersion: v1
 kind: Service
 metadata:
   name: {{ include "redpanda.fullname" . }}
   namespace: {{ .Release.Namespace | quote }}
   labels:
+    prometheus/scrape: "true"
```
Comment (joejulian) [inline charts/redpanda/templates/service.internal.yaml:27]: should this be gated on monitoring.enabled?

File: charts/redpanda/templates/service.internal.yaml [hunk 2]
```diff
@@ -34,3 +35,8 @@ spec:
   selector:
     app.kubernetes.io/name: {{ template "redpanda.name" . }}
     app.kubernetes.io/instance: {{ .Release.Name | quote }}
+  ports:
+    - name: admin
+      protocol: TCP
+      targetPort: {{ .Values.listeners.admin.port }}
+      port: {{ .Values.listeners.admin.port }}
```
Comment (joejulian) [inline charts/redpanda/templates/service.internal.yaml:42]: Should this also be gated?

File: charts/redpanda/templates/servicemonitor.yaml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+metadata:
+  name: {{ include "redpanda.fullname" . }}
+  namespace: {{ .Release.Namespace | quote }}
+  labels:
+{{- with include "full.labels" . }}
+  {{- . | nindent 4 }}
+{{- end }}
+spec:
+  endpoints:
+  - interval: 30s
```
Comment (joejulian) [inline charts/redpanda/templates/servicemonitor.yaml:31]: Should this use a value under monitoring?

File: charts/redpanda/templates/service.internal.yaml (hunk 1 above)
Comment (joejulian) [reply charts/redpanda/templates/service.internal.yaml:27]: Is this even needed if you create a ServiceMonitor? iirc, you don't need both.

Comment (prune998) [reply charts/redpanda/templates/service.internal.yaml:27]: I actually set this specific label so we can differenciate between this service and the others services from the chart, but sure I can gate it

Comment (prune998) [reply charts/redpanda/templates/service.internal.yaml]: The `serviceMonitor` needs to select a `Service` based on labels. So I need to be able to pick the right one between the `service.internal`, `service.loadbalancer` and `service.nodeport` as they all show the same list of labels...

Comment (joejulian) [reply charts/redpanda/templates/service.internal.yaml]: Got it! Thanks.

**Review by joejulian — APPROVED** [review]

File: charts/redpanda/templates/service.internal.yaml [hunk 4]
```diff
@@ -16,14 +16,15 @@ limitations under the License.
...
-# application
+# the stateful set and allow the serviceMonitor to target the pods.
+# This service should not be used by any client application
 apiVersion: v1
 kind: Service
 metadata:
   name: {{ include "redpanda.fullname" . }}
   namespace: {{ .Release.Namespace | quote }}
   labels:
+    monitoring/enabled: "{{ .Values.monitoring.enabled }}"
```
Comment (joejulian) [inline charts/redpanda/templates/service.internal.yaml:27]: To distinguish labels that are specific to our use case, it's good practice to add a domain.

```suggestion
    monitoring.redpanda.com/enabled: "{{ .Values.monitoring.enabled }}"
```

File: charts/redpanda/templates/servicemonitor.yaml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  {{- . | nindent 4 }}
+{{- end }}
+spec:
+  endpoints:
+  - interval: {{ .Values.monitoring.scrapeInterval }}
+    path: /public_metrics
+    targetPort: admin
+  selector:
+    matchLabels:
+      monitoring/enabled: "true"
```
Comment (joejulian) [inline charts/redpanda/templates/servicemonitor.yaml:36]:
```suggestion
      monitoring.redpanda.com/enabled: "true"
```

**Review by joejulian — COMMENTED** [review]
Will need a minor version bump in Chart.yaml as this is a new feature. That will fix the linter.

**Comment by prune998** [comment]
not sure what to do with the failed check... ?

**Comment by joejulian** [comment]
It's not a required check. It fails if there are any changes to the values files to ensure no accidental changes are merged.

Looks good, thanks!

**Pull request merged**